rails/database-migrations.md

```diff
@@ -0,0 +1,164 @@
+
+Now, another developer rebases his feature branch onto master and attempts to run the pending migration.
+
+Failure. To be specific, a `NoMethodError: undefined method` failure because the association the migration depended on isn't defined on the model anymore, so the migration blows up. 😩
+
+You can avoid this problem by following a simple rule:
+
+> Migrations may only use code defined within their own class.
+
+If you're migrating data and need helper methods or ActiveRecord associations, define them directly in the migration class itself. See the **Use Cases** section below for an example of how to properly use ActiveRecord associations in a migration.
+
+
+### Do not use migrations to create new data
+Use `seeds.rb` or a rake task to populate necessary app data. Otherwise running `db:schema:load` will leave the app's database without the needed data.
+
+### Always test your migrations
+Always run the migration, roll it back, and re-run it (`rails db:migrate:redo` is handy for this). This prevents accidentally writing irreversible migrations because you forgot to specify all the needed details the `change` method needs to calculate the reverse of the operation you're running.
+
+If a migration by its nature cannot be rolled back, be explicit about it by raising `IrreversibleMigration` in the `down` method.
+
+Other common gotchas to test for:
+- The migration won't run on a fresh database because there was something in the local dev db it was relying on.
+- Use realistic production sized data to check for speed.
+
+## Use Cases
+
+### Use ActiveRecord associations in a migration
+
+Existing model:
+
+```ruby
+class Tag
+  belongs_to :blog_post
+end
+
+class BlogPost
+  has_many :tags
+end
+```
+
+You decide for performance reasons to move the tags into an array column on the `blog_posts` table instead of having them in a separate `tags` table.
+
+Write the migration this way:
+
+```ruby
+class MoveTags < ActiveRecord::Migration[5.2]
+  # Inherit from ActiveRecord::Base instead of ApplicationRecord since the
+  # ApplicationRecord is subject to changes as the app changes.
+  class Tag < ActiveRecord::Base; end # rubocop:disable Rails/ApplicationRecord
+
+  class BlogPost < ActiveRecord::Base # rubocop:disable Rails/ApplicationRecord
+    has_many :tags
+  end
+
+  def up
+    add_column :blog_posts, :categories, :string, array: true
+
+    say("Moving #{Tag.count} tags to #{BlogPost.count} blog posts")
+
+    # Use find_each to minimize memory usage when dealing with many records
+    BlogPost.find_each(batch_size: 100) do |blog_post|
+      tags = blog_post.tags.pluck(:value)
+
+      # You can never have too much logging when migrating data!
+      say("Moving tags #{tags} to BlogPost #{blog_post.id}")
+
+      blog_post.update!(categories: tags)
```

Comment:
don't you need to reset column information before doing this?

Reply:
@johncalvinyoung The class is being redefined in this class so it doesn't need reset as a result.